### ISSUE TITLE
[QUESTION] pytorch distributed with openmpi

### ISSUE THREAD

**Issue by fwang2** [open-ce/open-ce#376]
hi there -

Based on the doc `open-ce build env`, the default is to be built with `openmpi` as default. However, the build I have generated doesn't seem to have it for `pytorch`. The follow snippet will fail. Is there any particular option I need to provide or change for `pytorch-feedstock` to enable it? TIA.

```
import torch.distributed as dist
dist.init_process_group(backend='mpi')
...
python3.7/site-packages/torch/distributed/distributed_c10d.py", line 496, in _new_process_group_helper
    "Distributed package doesn't have MPI built in."
RuntimeError: Distributed package doesn't have MPI built in. MPI is only included if you build PyTorch from source on a host that has MPI installed.
>>>

```

**Comment by jayfurmanek** [comment]
We don’t enable the MPI option in pytorch for a few reasons:
- it’s not well supported
- Both the pytorch native gloo and horovod perform much better.

Can you use one of those two?
If not it shouldn’t be hard to get it enabled in the pytorch recipe.

Are you using GPUs?

**Comment by fwang2** [comment]
point taken. thanks.
I am checking this option for two reasons 1) the readme suggested openmpi as default, which lead me think this is built-in; and for that matter, I also tried "gloo" and "nccl" backend, none works. 2) the snippet above is often the simplest and plain, and user tried them on our system to no avail. yes, we do use GPU.

**Comment by bnemanich** [comment]
In order to enable MPI within the PyTorch feedstock, you will need to:
* set `export USE_MPI=1` in `recipe/build.sh`
* Add `openmpi` as a build and runtime dependency within `recipe/meta.yaml`.

However, as @jayfurmanek mentioned, for PyTorch it is generally better to either use one of the more well supported backends, such as NCCL or GLOO, or use Horovod for distributed training.

**Comment by jayfurmanek** [comment]
You may also have to edit the `meta.yaml` to include openmpi as a build requirement so that its present for the Pytorch build to detect.

But...If the `nccl` backend doesn't work then perhaps there is something else going on here with your set up.
When you saw the `nccl` backend doesn't work - what have you tried?

**Comment by fwang2** [comment]
@jayfurmanek  both `nccl` and `gloo` actually work; I missed something earlier, sorry.

In the spirit of giving it a last shot, I did modify both `USE_MPI=1` and `meta.yaml` in pytorch feedstock, and having openmpi as a run dependency:
```
 run:
    - python {{ python }}
    - numpy {{ numpy }}
    - cffi {{ cffi }}
    - openmpi {{ openmpi }}
    ... 
```

I don't know the specific rules on the formatting part ... though. Anyway, the change seems mess something up as seen in the following output.


```
(base) [builder@303b8230a28c ~]$ open-ce build env --build_types=cuda --cuda_version 10.2 --mpi_types openmpi open-ce-environments/envs/pytorch-env.yaml
/opt/conda/lib/python3.8/site-packages/conda_build/environ.py:447: UserWarning: The environment variable 'CUDA_HOME' is being passed through with value '/usr/local/cuda'.  If you are splitting build and test phases with --no-test, please ensure that this value is also set similarly at test time.
  warnings.warn(
[OPEN-CE-ERROR-10] Error creating Build Tree
[OPEN-CE-ERROR-29] Failure getting remote dependencies for the following packages:
{({'leveldb'} : None), ({'lmdb 0.9.22.*'} : None), ({'requests 2.22'} : None), ({'libgfortran-ng 7.3.*'} : None), ({'cython 0.29'} : None), ({'pip 19*'} : None), ({'six 1.15.0'} : None), ({'gcc_linux-ppc64le 7.3.*'} : None), ({'libprotobuf 3.9.*'} : None), ({'ffmpeg'} : None), ({'numpy 1.19.*'} : None), ({'setuptools 41.*'} : None), ({'cffi 1.14.*'} : None), ({'decorator'} : None), ({'gfortran_linux-ppc64le 7.3.*'} : None), ({'gxx_linux-ppc64le 7.3.*'} : None), ({'patch'} : None), ({'ninja 1.9.*'} : None), ({'pyyaml 5.*'} : None), ({'pytest-runner 5.*'} : None), ({'python 3.7'} : None), ({'cmake 3.14.*'} : None), ({'protobuf 3.9.*'} : None), ({'libgcc-ng'} : None), ({'scipy 1.4.*'} : None), ({'networkx 2.3.*'} : None), ({'libtool 2.4.6.*'} : None), ({'autoconf 2.69.*'} : None), ({'cudatoolkit  10.2'} : None), ({'wheel 0.35*'} : None), ({'pkg-config 0.29.2'} : None), ({'typing_extensions 3.7.4.*'} : None), ({'pillow 8.1.*'} : None), ({'libpng'} : None), ({'future 0.18.*'} : None), ({'jpeg 9b'} : None), ({'make 4.2.1.*'} : None), ({'git >=2.15'} : None), ({'pywget'} : None), ({'automake 1.16.*'} : None), ({'click 7.0.*'} : None), ({'patchelf'} : None), ({'openblas 0.3.6.*'} : None), ({'openblas-devel 0.3.6.*'} : None)}
Error:
[OPEN-CE-ERROR-28] Conda Package Info Failed.
Command:
conda search --json  "openmpi 3.1.3.*" --info
Output:
{
  "caused_by": "None",
  "channel_urls": [
    "https://repo.anaconda.com/pkgs/main/linux-ppc64le",
    "https://repo.anaconda.com/pkgs/main/noarch",
    "https://repo.anaconda.com/pkgs/r/linux-ppc64le",
    "https://repo.anaconda.com/pkgs/r/noarch"
  ],
  "channels_formatted": "  - https://repo.anaconda.com/pkgs/main/linux-ppc64le\n  - https://repo.anaconda.com/pkgs/main/noarch\n  - https://repo.anaconda.com/pkgs/r/linux-ppc64le\n  - https://repo.anaconda.com/pkgs/r/noarch",
  "error": "PackagesNotFoundError: The following packages are not available from current channels:\n\n  - openmpi=3.1.3\n\nCurrent channels:\n\n  - https://repo.anaconda.com/pkgs/main/linux-ppc64le\n  - https://repo.anaconda.com/pkgs/main/noarch\n  - https://repo.anaconda.com/pkgs/r/linux-ppc64le\n  - https://repo.anaconda.com/pkgs/r/noarch\n\nTo search for alternate channels that may provide the conda package you're\nlooking for, navigate to\n\n    https://anaconda.org\n\nand use the search bar at the top of the page.\n",
  "exception_name": "PackagesNotFoundError",
  "exception_type": "<class 'conda.exceptions.PackagesNotFoundError'>",
  "message": "The following packages are not available from current channels:\n\n  - openmpi=3.1.3\n\nCurrent channels:\n\n  - https://repo.anaconda.com/pkgs/main/linux-ppc64le\n  - https://repo.anaconda.com/pkgs/main/noarch\n  - https://repo.anaconda.com/pkgs/r/linux-ppc64le\n  - https://repo.anaconda.com/pkgs/r/noarch\n\nTo search for alternate channels that may provide the conda package you're\nlooking for, navigate to\n\n    https://anaconda.org\n\nand use the search bar at the top of the page.\n",
  "packages": [
    "openmpi=3.1.3"
  ],
  "packages_formatted": "  - openmpi=3.1.3"
}
```

**Comment by jayfurmanek** [comment]
ah, yeah you also have to add openmpi to that `pytorch-env.yaml` file.
